### PR TITLE
patron_transaction_events: create new resource :: jsonschema and mapping. 

### PR DESCRIPTION
**Pull request by zannkukai**
* Creation of 'patron_transaction_events'.

Co-Authored-by: Renaud Michotte <renaud.michotte@gmail.com>
Co-Authored-by: Aly Badr <aly.badr@rero.ch>

## Why are you opening this PR?

- Which task/US does it implement?
https://tree.taiga.io/project/rero21-reroils/task/1250?kanban-status=1224894

## How to test?

Other pull requests will follow, only code review is needed.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?

### CODE REVIEW

File: rero_ils/modules/patron_transaction_events/jsonschemas/patron_transaction_events/patron_transaction_event-v0.0.1.json [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+      "minLength": 1
+    },
+    "parent": {
+      "title": "Parent patron transaction",
+      "type": "object",
+      "properties": {
+        "$ref": {
+          "title": "Patron transaction URI",
+          "type": "string",
+          "pattern": "^https://ils.rero.ch/api/patron_transaction/.*?$"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/jsonschemas/patron_transaction_events/patron_transaction_event-v0.0.1.json]: `patron_transactions` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    "creation_date": "2020-01-24T00:00:00",
+    "amount": 2,
+    "type": "fee",
+    "subtype": "overdue",
+    "status": "closed",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/1222"
+    },
+    "parent": {
+      "$ref": "https://ils.rero.ch/api/patron_transaction/1"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: `patron_transactions` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    "creation_date": "2020-01-25T00:00:00",
+    "amount": 0.5,
+    "type": "fee",
+    "subtype": "overdue",
+    "status": "open",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/1222"
+    },
+    "parent": {
+      "$ref": "https://ils.rero.ch/api/patron_transaction/1"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: `patron_transactions` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    "subtype": "cash",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/1222"
+    },
+    "operator": {
+      "$ref": "https://ils.rero.ch/api/patrons/badr.aly"
+    },
+    "note": "Pay with low-value coins... dummy student",
+    "parent": {
+      "$ref": "https://ils.rero.ch/api/patron_transaction/1"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: `patron_transactions` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    "type": "payment",
+    "subtype": "paypal",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/1222"
+    },
+    "operator": {
+      "$ref": "https://ils.rero.ch/api/patrons/badr.aly"
+    },
+    "parent": {
+      "$ref": "https://ils.rero.ch/api/patron_transaction/1"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: `patron_transactions` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    "pid": "3",
+    "creation_date": "2020-01-25T10:00:00",
+    "amount": 1,
+    "type": "payment",
+    "subtype": "cash",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/1222"
+    },
+    "operator": {
+      "$ref": "https://ils.rero.ch/api/patrons/badr.aly"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: Why `badr.aly` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    "pid": "4",
+    "creation_date": "2020-01-25T10:00:20",
+    "amount": 1,
+    "type": "payment",
+    "subtype": "paypal",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/1222"
+    },
+    "operator": {
+      "$ref": "https://ils.rero.ch/api/patrons/badr.aly"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: Why `badr.aly` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  {
+    "$schema": "https://ils.rero.ch/schema/patron_transaction_events/patron_transaction_event-v0.0.1.json",
+    "pid": "5",
+    "creation_date": "2020-01-25T10:10:00",
+    "type": "dispute",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/1222"
+    },
+    "operator": {
+      "$ref": "https://ils.rero.ch/api/patrons/badr.aly"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: Why `badr.aly` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    },
+    "note": "Student doesn't want to pay for overdue 2020-01-25",
+    "parent": {
+      "$ref": "https://ils.rero.ch/api/patron_transaction/1"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: `patron_transactions` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  {
+    "$schema": "https://ils.rero.ch/schema/patron_transaction_events/patron_transaction_event-v0.0.1.json",
+    "pid": "6",
+    "creation_date": "2020-01-31T16:22:32",
+    "type": "resolved",
+    "location": {
+      "$ref": "https://ils.rero.ch/api/locations/4"
+    },
+    "operator": {
+      "$ref": "https://ils.rero.ch/api/patrons/god.master"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: Why `god.master` ?

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    },
+    "note": "Waive for the dispute",
+    "amount": 0.5,
+    "parent": {
+      "$ref": "https://ils.rero.ch/api/patron_transaction/1"
```
Comment (Garfield-fr) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: `patron_transactions` ?

**Comment by Garfield-fr** [comment]
Don't forget to extract the messages

File: rero_ils/modules/patron_transaction_events/api.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+class PatronTransactionEvent(IlsRecord):
+    """PatronTransactionEvent class."""
+
+    minter = patron_transaction_event_id_minter
+    fetcher = patron_transaction_event_id_fetcher
+    provider = PatronTransactionEventProvider
+
+    @classmethod
+    def create(cls, data, id_=None, delete_pid=False,
```
Comment (BadrAly) [inline rero_ils/modules/patron_transaction_events/api.py]: This function not needed, can be deleted

File: rero_ils/modules/patron_transaction_events/api.py [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+               dbcommit=False, reindex=False, **kwargs):
+        """Create patron_transaction_event record."""
+        record = super(PatronTransactionEvent, cls).create(
+            data, id_, delete_pid, dbcommit, reindex, **kwargs)
+        return record
+
+    def update(self, data, dbcommit=False, reindex=False):
```
Comment (BadrAly) [inline rero_ils/modules/patron_transaction_events/api.py]: This function not needed, can be deleted

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json (hunk 6 above)
Comment (zannkukai) [reply rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: That's just a sample data file to explain all events occurred for an `overdue` with "dummy" reference within.
This file will be removed in next PR

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json (hunk 10 above)
Comment (zannkukai) [reply rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: as "badr.aly" --> sample data

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json [hunk 14]
```diff
@@ -0,0 +1,100 @@
+[
+  {
+    "$schema": "https://ils.rero.ch/schema/patron_transaction_events/patron_transaction_event-v0.0.1.json",
+    "pid": "1",
```
Comment (BadrAly) [inline rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: instead of `pid=1` move to `pid=ptre1`

**Comment by BadrAly** [comment]
I think we will move the data from  `rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json` to `rero_ils/data/patron_transaction.json`

File: rero_ils/modules/patron_transaction_events/templates/rero_ils/detailed_view_patron_transaction_events.html [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+  You should have received a copy of the GNU Affero General Public License
+  along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#}
+
+{%- extends config.RECORDS_UI_BASE_TEMPLATE %}
+
+{%- block page_body %}
+
+{%- endblock %}
```
Comment (lauren-d) [inline rero_ils/modules/patron_transaction_events/templates/rero_ils/detailed_view_patron_transaction_events.html]: necessary?

File: rero_ils/modules/patron_transaction_events/views.py [hunk 16]
```diff
@@ -0,0 +1,29 @@
...
+
+from __future__ import absolute_import, print_function
+
+from flask import Blueprint
+
+blueprint = Blueprint(
+    'patron_transaction_events',
+    __name__,
+    template_folder='templates',
+    static_folder='static',
```
Comment (lauren-d) [inline rero_ils/modules/patron_transaction_events/views.py]: remove views.py if not needed

File: rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json (hunk 14 above)
Comment (zannkukai) [reply rero_ils/modules/patron_transaction_events/sample_data/sample-scenario-overdue1.json]: According to other resources in `data` folder, the pid are always just a digit.
'ptre1' should be used into fixtures

**Pull request merged**